### ISSUE TITLE
QuadratureSpace support for element faces?

### ISSUE THREAD

**Issue by artv3** [mfem/mfem#2863]
Folks, I think we may be missing QuadratureSpace support for element face quadrature rules. Could we introduce a QuadratureFaceSpace?

**Comment by artv3** [comment]
Just to add more detail, my use-case for the volumetric terms is as follows:

```
QuadratureSpace qspace(pmesh,order)    
QuadratureFunction q_fun(qspace, dim); 
VectorQuadratureFunctionCoefficient q_coefficient(q_fun);
```

At the moment qspace will intialize a quadrature rule with quad_1D^{mesh_dim} values, but for the face terms we need it to be mesh_dim -1, and I don't see how this can be done with the current MFEM tools.

**Comment by YohannDudouit** [comment]
@pazner did this in the libCEED integration to avoid this limitation:
```c++
   // The x-coordinates of the first `nqpts` points of the integration rule are
   // the points of the corresponding 1D rule. We also scale the weights
   // accordingly.
   double w_sum = 0.0;
   for (int i = 0; i < nqpts; i++)
   {
      const mfem::IntegrationPoint &ip = ir.IntPoint(i);
      qX(i) = ip.x;
      qW(i) = ip.weight;
      w_sum += ip.weight;
   }
   qW *= 1.0/w_sum;
```
But I agree that we should find a convenient way to access this for users.

**Comment by v-dobrev** [comment]
We should try to address this as part of #1656. It may be more convenient to allow the same class, `QuadratureSpace`, to support both volume and face quadrature spaces -- that way evaluation of coefficients on volume and face quadrature spaces can use the same interface.

**Comment by pazner** [comment]
I looked into what it would take to implement this feature. I think we need to make a design decision, there are two main possibilities. What are everyone's thoughts regarding options 1️⃣ and 2️⃣ below?

### 1️⃣ Separate classes for face and element spaces and functions

**Pros:**

* Better use of type system to ensure correctness at compile-time
* More consistent and clearer interface
* Much less risk of writing code that e.g. assumes that a `QuadratureFunction` is defined on elements, and then crashes because it's really defined on faces.

**Cons:**

* Might result in more duplication (will we need a `FaceQuadratureFunctionCoefficient` in addition to a `QuadratureFunctionCoefficient`?)

### 2️⃣  Reuse `QuadratureSpace` and `QuadratureFunction` for faces

**Pros:**
* May result in less code duplication 

**Cons:**
* Weaker use of type system for compile-time correctness checks
* Some of the member functions don't make sense for faces (e.g. will `QuadratureFunction::GetElementValues` actually get face values, or will it just give a run-time error?)

**Comment by tzanio** [comment]
Thanks @pazner. I vote for 1️⃣.

**Comment by artv3** [comment]
1. looks good to me too

**Comment by YohannDudouit** [comment]
I think 1️⃣  is the safest and cleanest path.

**Comment by v-dobrev** [comment]
This issue will be addressed as part of the unification of #1656 and #2891, so I'll close it.
